### PR TITLE
Add category model and functionality for it

### PR DESCRIPTION
**Pull request by Den-moroz**

### CODE REVIEW

File: src/main/java/core/basesyntax/bookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    @GetMapping
+    @Operation(summary = "Get all categories", description = "Get a list of all category, "
+            + "pagination and sorting are also included")
+    public List<CategoryDto> getAll(Pageable pageable) {
+        return categoryService.findAll(pageable);
+    }
+
+    @GetMapping(path = "/{id}")
+    public CategoryDto getCategoryById(@PathVariable Long id) {
```
Comment (satori11111) [inline src/main/java/core/basesyntax/bookstore/controller/CategoryController.java:55]: @Operation documentation missed

File: src/main/java/core/basesyntax/bookstore/dto/category/CreateCategoryRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+import org.hibernate.validator.constraints.Length;
+
+@Data
+public class CreateCategoryRequestDto {
+    @NotNull
+    @Length(min = 1, max = 25)
+    private String name;
+    private String description;
```
Comment (satori11111) [inline src/main/java/core/basesyntax/bookstore/dto/category/CreateCategoryRequestDto.java:13]: min Lenght description?

File: src/main/resources/db/changelog/changes/05-create-book-category-table.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+              - column:
+                  name: book_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+              - column:
+                  name: category_id
+                  type: bigint
+                  constraints:
+                    nullable: false
```
Comment (satori11111) [inline src/main/resources/db/changelog/changes/05-create-book-category-table.yaml]: add newLine and what if you insert book with non-existing categories? you need to add foreignKey and primaryKey

**Review by satori11111 — CHANGES_REQUESTED** [review]

File: src/main/resources/db/changelog/changes/05-create-book-category-table.yaml (hunk 3 above)
Comment (Den-moroz) [reply src/main/resources/db/changelog/changes/05-create-book-category-table.yaml]: i fix it in next repo and other liquibase files

**Review by satori11111 — APPROVED** [review]

File: src/main/java/core/basesyntax/bookstore/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Category management", description = "Endpoints for managing category")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/categories")
```
Comment (vovasalyha) [inline src/main/java/core/basesyntax/bookstore/controller/CategoryController.java:32]:
```suggestion
@RequestMapping(value = "/api/categories")

```

File: src/main/java/core/basesyntax/bookstore/model/Book.java [hunk 5]
```diff
@@ -30,6 +37,13 @@ public class Book {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(name = "book_category",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id")
+    )
+    @ToString.Exclude
```
Comment (vovasalyha) [inline src/main/java/core/basesyntax/bookstore/model/Book.java:46]: Exclude it from equals and hashcode too.

File: src/main/java/core/basesyntax/bookstore/model/Category.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Data;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Data
+@Entity
+@SQLDelete(sql = "UPDATE category SET is_deleted = true WHERE id=?")
```
Comment (vovasalyha) [inline src/main/java/core/basesyntax/bookstore/model/Category.java]:
```suggestion
@SQLDelete(sql = "UPDATE category SET is_deleted = true WHERE id= ?")
```

File: src/main/java/core/basesyntax/bookstore/model/Category.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+@Where(clause = "is_deleted=false")
```
Comment (vovasalyha) [inline src/main/java/core/basesyntax/bookstore/model/Category.java]:
```suggestion
@Where(clause = "is_deleted = false")
```

File: src/main/resources/db/changelog/changes/04-create-category-table.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+        - addUniqueConstraint:
+            tableName: category
+            columnNames: name
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/04-create-category-table.yaml]: Empty line.

File: src/main/resources/db/changelog/changes/05-create-book-category-table.yaml (hunk 3 above)
Comment (vovasalyha) [reply src/main/resources/db/changelog/changes/05-create-book-category-table.yaml]: Why not fix it in this PR? You are trying to merge broken code

File: src/main/java/core/basesyntax/bookstore/repository/book/BookRepository.java [hunk 9]
```diff
@@ -1,9 +1,25 @@
...
+import java.util.List;
+import java.util.Optional;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
 
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
```
Comment (vovasalyha) [inline src/main/java/core/basesyntax/bookstore/repository/book/BookRepository.java]: Do you really need to fetch categories for all methods? And if you do so, specify in method name that your method fetches categories too, similar to what you named it in `findAllWithCategories`

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Nice. Let's improve it a bit.

File: src/main/resources/db/changelog/changes/05-create-category-table.yaml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+        - addUniqueConstraint:
+            tableName: category
+            columnNames: name
```
Comment (boroda4436) [inline src/main/resources/db/changelog/changes/05-create-category-table.yaml]: empty line)

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Please resolve merge conflicts.

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**